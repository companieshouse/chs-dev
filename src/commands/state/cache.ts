import { Args, Command, Config, Flags } from "@oclif/core";
import { StateManager } from "../../state/state-manager.js";
import ChsDevConfig from "../../model/Config.js";
import loadConfig from "../../helpers/config-loader.js";
import { DockerCompose } from "../../run/docker-compose.js";
import { confirm } from "../../helpers/user-input.js";

export default class Cache extends Command {

    static description = "Cache the state of chs-dev into a saved file";

    static args = {
        name: Args.string({
            required: false,
            description: "Name of the cache"
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        const logger = {
            log: (msg: string) => this.log(msg)
        };
        this.chsDevConfig = loadConfig();
        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);

    }

    async run (): Promise<any> {
        const { args, flags } = await this.parse(Cache);
        const timestamp: string = new Date(Date.now()).toISOString();
        const cacheName: string = args.name || timestamp;

        const handlePrompt = async (): Promise<boolean> => {
            return await confirm(`save chs-dev state?`);
        };

        if (await handlePrompt()) {
            //
        }
    }

}
