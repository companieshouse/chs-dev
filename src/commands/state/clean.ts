import { Command, Config, Flags } from "@oclif/core";
import { StateManager } from "../../state/state-manager.js";
import ChsDevConfig from "../../model/Config.js";
import loadConfig from "../../helpers/config-loader.js";
import { DockerCompose } from "../../run/docker-compose.js";
import { confirm } from "../../helpers/user-input.js";

export default class Clean extends Command {

    static description = "Wipes the state of chs-dev and revert to a default state";

    static flags = {
        purge: Flags.boolean({
            char: "p",
            description: "Wipe state, volumes and images",
            default: false
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        const logger = {
            log: (msg: string) => this.log(msg)
        };
        this.chsDevConfig = loadConfig();
        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);

    }

    async run (): Promise<any> {
        const { flags: { purge } } = await this.parse(Clean);

        const handlePrompt = async (): Promise<boolean> => {
            return await confirm(`This will reset your chs-dev state and delete all unused docker volumes${purge ? " including images." : "."} Do you want to proceed?`);
        };

        if (await handlePrompt()) {
            const containerStatus = this.dockerCompose.getServiceStatuses();
            if (containerStatus !== undefined) {
                this.error("Ensure all containers are stopped. Run: '$ chs-dev down'");
            }
            this.stateManager.cleanState();

            await this.config.runHook("generate-runnable-docker-compose", {});

            this.dockerCompose.prune("volume");
            if (purge) {
                this.dockerCompose.prune("image");
            }

        }
    }

}
