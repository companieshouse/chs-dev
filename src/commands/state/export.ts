import { Args, Command, Config } from "@oclif/core";
import { existsSync, mkdirSync } from "fs";
import { join } from "path";
import yaml from "yaml";
import loadConfig from "../../helpers/config-loader.js";
import { getGeneratedDockerComposeFile } from "../../helpers/docker-compose-file.js";
import { writeContentToFile } from "../../helpers/file-utils.js";
import { hashAlgorithm } from "../../helpers/index.js";
import { confirm } from "../../helpers/user-input.js";
import ChsDevConfig from "../../model/Config.js";
import { StateManager } from "../../state/state-manager.js";

type StateCache = {
    state: {
        snapshot: Record<string, any>;
        hash: string
    }
    dockerCompose: {
        snapshot: Record<string, any>;
        hash: string;
    }
};
const EXPORT_STATE_DIR = ".exported_state_cache";

export default class Export extends Command {
    static description = "Export current state of chs-dev into a saved file";

    static args = {
        name: Args.string({
            required: true,
            description: "Name of the exported file"
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly exportedStateDir: string;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        this.chsDevConfig = loadConfig();
        this.exportedStateDir = join(this.chsDevConfig.projectPath, EXPORT_STATE_DIR);

        if (!existsSync(this.exportedStateDir)) {
            mkdirSync(this.exportedStateDir, { recursive: true });
        }

        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
    }

    async run (): Promise<void> {
        const { args: { name } } = await this.parse(Export);

        if (await confirm(`Export current state of chs-dev to file named: '${name}'?`)) {
            this.exportState(name);
        }

    }

    private exportState (exportFileName: string): void {
        const exportedFilenamePath = join(this.exportedStateDir, `${exportFileName}.yaml`);
        writeContentToFile(this.getExportData(), exportedFilenamePath);

        this.log(`Exported state destination: ${exportedFilenamePath}`);
    }

    private getExportData (): StateCache {
        const stateSnapshot = this.stateManager.snapshot;
        const dockerComposeSnapshot = getGeneratedDockerComposeFile(this.chsDevConfig.projectPath);
        return {
            state: {
                hash: hashAlgorithm(yaml.stringify(stateSnapshot)),
                snapshot: stateSnapshot
            },
            dockerCompose: {
                hash: hashAlgorithm(yaml.stringify(dockerComposeSnapshot)),
                snapshot: dockerComposeSnapshot
            }
        };

    }

}
