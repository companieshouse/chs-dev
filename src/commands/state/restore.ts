import { Args, Command, Config, Flags } from "@oclif/core";
import { existsSync } from "fs";
import { basename, join } from "path";
import yaml from "yaml";
import loadConfig from "../../helpers/config-loader.js";
import { readFileContent, writeContentToFile } from "../../helpers/file-utils.js";
import { hashAlgorithm } from "../../helpers/index.js";
import { confirm } from "../../helpers/user-input.js";
import ChsDevConfig from "../../model/Config.js";
import { DockerCompose } from "../../run/docker-compose.js";

type StateCache = {
    state: {
        snapshot: Record<string, any>;
        hash: string
    }
    dockerCompose: {
        snapshot: Record<string, any>;
        hash: string;
    }
};

export default class Restore extends Command {
    static description = "Restore and update the state files from saved cache or imported cache.";

    static args = {
        name: Args.string({
            required: false,
            description: "Name of the cache"
        })
    };

    static flags = {
        importCacheFrom: Flags.string({
            char: "i",
            description: "Path to the exported cache"
        })
    };

    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;
    private readonly stateCacheFile: string;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        this.chsDevConfig = loadConfig();
        const cacheDir = config.cacheDir;
        this.stateCacheFile = join(cacheDir, `${basename(this.chsDevConfig.projectName)}.state.yaml`);

        const logger = { log: (msg: string) => this.log(msg) };

        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);
    }

    async run (): Promise<void> {
        const { args, flags: { importCacheFrom } } = await this.parse(Restore);

        const containerStatus = this.dockerCompose.getServiceStatuses();
        if (containerStatus !== undefined) {
            this.error("Ensure all containers are stopped. Run: '$ chs-dev down'");
        }

        const cacheName = importCacheFrom ? "import-cache" : args.name;

        if (!cacheName || typeof cacheName === "undefined") {
            this.error("Please provide a valid cache name or use the --importCacheFrom or -i flag to restore from an imported cache.");
        }
        console.log(`Restoring state from cache: ${cacheName}`);
        if (importCacheFrom) {
            if (await this.handlePrompt(cacheName)) {
                this.restoreFromImport(importCacheFrom);
            }
        } else {
            if (await this.handlePrompt(cacheName)) {
                this.restoreFromSavedCache(cacheName);
            }
        }

    }

    private restoreFromSavedCache (cacheName: string): void {
        const cacheData = this.loadSavedCacheData();
        const namedCacheData = cacheData[cacheName];
        if (!namedCacheData) {
            this.error(`Cache with name '${cacheName}' not found.`);
        }
        const data = this.verifyCacheAuthencity(namedCacheData, (stateCache) => stateCache);

        this.restoreState(data);
        this.log(`Restored state from saved cache '${cacheName}'`);
    }

    private restoreFromImport (importCachePath: string): void {
        if (!existsSync(importCachePath)) {
            this.error(`Import cache file does not exist in location: ${importCachePath}`);
        }

        const cacheData = readFileContent(importCachePath) as StateCache;

        if (!cacheData || typeof cacheData !== "object") {
            this.error(`Invalid cache data in imported file: ${importCachePath}`);
        }

        const data = this.verifyCacheAuthencity(cacheData, (stateCache) => stateCache);

        this.restoreState(data);
        this.log(`Restored state from imported cache.`);
    }

    private verifyCacheAuthencity<T> (cacheData: StateCache, cacheSupplier: (stateCache: StateCache) => StateCache): StateCache {

        if (cacheData.state.hash === hashAlgorithm(yaml.stringify(cacheData.state.snapshot)) &&
            cacheData.dockerCompose.hash === hashAlgorithm(yaml.stringify(cacheData.dockerCompose.snapshot))) {
            return cacheSupplier(cacheData);
        }
        this.error("Cache data has been corrupted or touched.");
    }

    private restoreState ({ state, dockerCompose }: StateCache): void {
        // Restore state file
        const stateFilePath = join(this.chsDevConfig.projectPath, `.chs-dev.yaml`);
        writeContentToFile(state.snapshot, stateFilePath);

        // Restore docker-compose file
        const dockerComposeFilePath = join(this.chsDevConfig.projectPath, "docker-compose.yaml");
        writeContentToFile(dockerCompose.snapshot, dockerComposeFilePath);
    }

    private loadSavedCacheData (): Record<string, StateCache> {
        return readFileContent(this.stateCacheFile) as Record<string, StateCache> || {};
    }

    private async handlePrompt (cacheName: string): Promise<boolean> {
        const message = cacheName === "import-cache"
            ? `Restore from an imported cache?`
            : `Restore from a saved cache '${cacheName}'?`;
        return confirm(message);
    }
}
