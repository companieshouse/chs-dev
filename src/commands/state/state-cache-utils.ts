import { existsSync } from "fs";
import { join } from "path";
import yaml from "yaml";
import { readFileContent, writeContentToFile } from "../../helpers/file-utils.js";
import { hashAlgorithm } from "../../helpers/index.js";
import ChsDevConfig from "../../model/Config.js";
import { confirm } from "../../helpers/user-input.js";

type StateCache = {
    state: {
        snapshot: Record<string, any>;
        hash: string;
    };
    dockerCompose: {
        snapshot: Record<string, any>;
        hash: string;
    };
};
type CacheActions = "add" | "remove" | "wipe" | "import-file-cache" | "import-saved-cache";

export const EXPORT_STATE_DIR = "exported_state_cache";

export const handlePrompt = async (
    action:CacheActions,
    cacheName: string = "default-cache-name"
): Promise<boolean> => {
    const messages: Record<string, string> = {
        add: `This will save the cache as '${cacheName}' or overwrite if name already exists. Proceed?`,
        remove: `Do you want to delete the cache '${cacheName}'?`,
        wipe: "Do you want to delete all saved caches?",
        "import-file-cache": "Restore from an imported file cache?",
        "import-saved-cache": `Restore from a saved cache '${cacheName}'?`
    };
    return await confirm(messages[action]);
};

export const validateNameFormat = (name: string): void => {
    const regex = /^[a-zA-Z0-9_-]+$/;
    if (!regex.test(name)) {
        throw new Error(`Invalid cache name format: '${name}'. Only alphanumeric characters with underscores or hypens are allowed.`);
    }
};

export const validateCacheNameExists = (cacheData: Record<string, any>, cacheName: string): void => {
    if (!cacheData[cacheName]) {
        throw new Error(`Cache named ${cacheName} does not exist.`);
    }
};

export function verifyCacheAuthenticity (cacheData: StateCache): StateCache {
    if (
        cacheData.state.hash === hashAlgorithm(yaml.stringify(cacheData.state.snapshot)) &&
            cacheData.dockerCompose.hash === hashAlgorithm(yaml.stringify(cacheData.dockerCompose.snapshot))
    ) {
        return cacheData;
    }
    throw new Error("Cache data has been corrupted or touched.");
}

export function loadImportCache (importCachePath: string): StateCache {
    if (!existsSync(importCachePath)) {
        throw new Error(`Import cache file does not exist in location: ${importCachePath}`);
    }
    const cacheData = readFileContent(importCachePath) as StateCache;
    if (!cacheData || typeof cacheData !== "object") {
        throw new Error(`Invalid cache data in imported file: ${importCachePath}`);
    }
    return cacheData;
}

export function restoreStateFiles (config: ChsDevConfig, state: StateCache["state"], dockerCompose: StateCache["dockerCompose"]): void {
    const stateFilePath = join(config.projectPath, ".chs-dev.yaml");
    writeContentToFile(state.snapshot, stateFilePath);
    const dockerComposeFilePath = join(config.projectPath, "docker-compose.yaml");
    writeContentToFile(dockerCompose.snapshot, dockerComposeFilePath);
}
