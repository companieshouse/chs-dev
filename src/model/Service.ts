import { DependencyNode } from "./DependencyGraph.js";
import { OrNull } from "./OrNull.js";

/**
 * Represents a service within the development environment. A service is a
 * single deployable unit - typically container.
 */
export interface Service {
    /**
     * Meaningful name for service, unique within the environment
     */
    name: string;

    /**
     * Name of the module which the service resides
     */
    module: string;

    /**
     * Human readable description of the service
     */
    description?: OrNull<string>;

    /**
     * Path to the Docker Compose file for the service
     */
    source: string;

    /**
     * List of services names (exhaustive) that the service depends on.
     *
     * This will include all of the direct and indirect/transitive dependencies
     * the service has.
     */
    dependsOn: string[];

    /**
     * Tree of the dependencies for the service.
     *
     * This will include all of the direct and indirect/transitive dependencies
     * the service has.
     */
    dependencyTree?: DependencyNode;

    /**
     * The number of times this service is used by other services.
     *
     * This is the number of services that have this service in their `dependsOn` list and
     * is not the same as `numberOfDependencies` which counts the number of
     * services that the service depends on.
     */
    timesUsedByOtherServices?: number;
    /**
     * The number of services that depend on this service.
     *
     * This is the number of services that this service depends on either directly or
     * indirectly. This is not the same as `timesUsedByOtherServices` which counts
     * the number of time a service appears on the `dependsOn` list.
     */
    numberOfDependencies?: number;
    /**
     * Details about the code repository for the service (required for
     * development mode)
     */
    repository: OrNull<{ url: string; branch?: string | null; }>;

    /**
     * When supplied will specify a common Dockerfile to use to build the
     * service for development mode
     */
    builder: string;

    /**
     * Other pertinent information about the service which may vary
     * service to service.
     */
    metadata: Record<string, OrNull<string> | undefined>;
}

export default Service;
