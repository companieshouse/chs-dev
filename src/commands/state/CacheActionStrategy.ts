import yaml from "yaml";

import { getGeneratedDockerComposeFile } from "../../helpers/docker-compose-file.js";
import { hashAlgorithm } from "../../helpers/index.js";
import { existsSync, mkdirSync } from "fs";
import { join } from "path";
import { writeContentToFile } from "../../helpers/file-utils.js";
import { handlePrompt, validateCacheNameExists } from "./state-cache-utils.js";
import { StateManager } from "../../state/state-manager.js";

const EXPORT_STATE_DIR = ".exported_state_cache";

export interface AvailableCacheContext {
    logger: (message: string) => void;
  }

export interface WipeCacheContext {
    stateCacheFile: string;
    logger: (message: string) => void;
}

export interface RemoveCacheContext {
    stateCacheFile: string;
    logger: (message: string) => void;
}

export interface ExportCacheContext {
    projectPath: string;
    logger: (message: string) => void;
}

export interface AddCacheContext {
    stateManager: StateManager;
    stateCacheFile: string;
    projectPath: string;
    logger: (message: string) => void;
}
interface CacheActionStrategy<C> {
    execute(cache: C, cacheData: Record<string, any>, cacheName?: string): Promise<void> | void;
}

const handleExportCache = (context: ExportCacheContext, cacheData: Record<string, any>, cacheName: string): void => {
    const exportDir = join(context.projectPath, EXPORT_STATE_DIR);
    if (!existsSync(exportDir)) {
        mkdirSync(exportDir, { recursive: true });
    }

    const exportedFilenamePath = join(exportDir, `${cacheName}.yaml`);

    writeContentToFile(cacheData[cacheName], exportedFilenamePath);
    context.logger(`Exported cache ${cacheName} destination: '${exportedFilenamePath}'`);
};

export class AvailableCacheStrategy implements CacheActionStrategy<AvailableCacheContext> {
    execute (context: AvailableCacheContext, cacheData: Record<string, any>): void {
        const cacheNames = Object.keys(cacheData);
        if (cacheNames.length > 0) {
            context.logger("Available caches:");
            cacheNames.forEach((name: string) => context.logger(`- ${name}`));
        } else {
            context.logger("No cache available.");
        }
    }
}

export class WipeCacheStrategy implements CacheActionStrategy<WipeCacheContext> {
    async execute (context: WipeCacheContext, _cacheData: Record<string, any>): Promise<void> {
        const confirmed = await handlePrompt("wipe");
        if (confirmed) {
            writeContentToFile({}, context.stateCacheFile);
            context.logger("Wiped all caches");
        }
    }
}

export class RemoveCacheStrategy implements CacheActionStrategy<RemoveCacheContext> {
    async execute (context: RemoveCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        validateCacheNameExists(_cacheData, _cacheName);

        const confirmed = await handlePrompt("remove", _cacheName);
        if (confirmed) {
            delete _cacheData[_cacheName];
            writeContentToFile(_cacheData, context.stateCacheFile);
            context.logger(`Removed cache '${_cacheName}'`);

        }
    }
}

export class ExportCacheStrategy implements CacheActionStrategy<ExportCacheContext> {
    async execute (context: ExportCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        validateCacheNameExists(_cacheData, _cacheName);
        handleExportCache(context, _cacheData, _cacheName);

    }
}

export class AddCacheStrategy implements CacheActionStrategy<AddCacheContext> {
    async execute (context: AddCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        const confirmed = await handlePrompt("add", _cacheName);
        if (confirmed) {
            const stateSnapshot = context.stateManager.snapshot;
            const dockerComposeSnapshot = getGeneratedDockerComposeFile(context.projectPath);
            _cacheData[_cacheName] = {
                state: {
                    hash: hashAlgorithm(yaml.stringify(stateSnapshot)),
                    snapshot: stateSnapshot
                },
                dockerCompose: {
                    hash: hashAlgorithm(yaml.stringify(dockerComposeSnapshot)),
                    snapshot: dockerComposeSnapshot
                }
            };
            writeContentToFile(_cacheData, context.stateCacheFile);
            context.logger(`Saved cache as '${_cacheName}'`);
        }
    }
}
