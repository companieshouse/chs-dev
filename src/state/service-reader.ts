import { basename, dirname } from "path";
import Service from "../model/Service.js";
import yaml from "yaml";
import { readFileSync } from "fs";

interface DependencySpecification {
    condition: "service_started" | "service_healthy" | "service_complete";
    restart?: boolean;
    required?: boolean;
}

interface DependencySpecificationMap {
    [dependency_name: string]: DependencySpecification;
}

interface ServiceDefinition {
    labels: string[];
    depends_on?: string[] | DependencySpecificationMap;
    healthcheck?: Record<string, any>;
}

/**
 * Parses a docker-compose.yaml file and creates a Service from it
 * @param filePath Docker Compose config file
 * @returns Partially complete Service with only direct dependencies
 */
export const readServices: (filePath: string) => Partial<Service>[] = (filePath: string) => {
    const module = basename(dirname(filePath));
    const services = yaml.parse(readFileSync(filePath).toString()).services as { [serviceName: string]: ServiceDefinition };
    const source = filePath;

    return Object.entries(services)
        .map(([serviceName, service]) => readService(module, source, serviceName, service));
};

const metadataLabelMapping = {
    repoContext: "chs.local.repoContext",
    // The escaped \. indicates a partial label and so this is not a
    // useless escape
    // eslint-disable-next-line no-useless-escape
    ingressRoute: "traefik.http.routers\.",
    languageMajorVersion: "chs.local.builder.languageVersion",
    dockerfile: "chs.local.dockerfile",
    entrypoint: "chs.local.entrypoint",
    buildOutputDir: "chs.local.builder.outputDir",
    secretsRequired: "chs.local.builder.requiresSecrets",
    repositoryRequired: "chs.local.repositoryRequired",
    builderUseServiceDockerFile: "chs.local.builder.useServiceDockerfile"
};

const readService: (module: string, source: string, serviceName: string, service: ServiceDefinition) => Partial<Service> = (module, source, serviceName, service) => ({
    name: serviceName,
    module,
    source,
    description: findLabel(service.labels, "chs.description"),
    dependsOn: parseDependsOn(service.depends_on),
    repository: parseRepository(service.labels),
    builder: findLabel(service.labels, "chs.local.builder") || "",
    metadata: {
        ...readLabels(service.labels),
        healthcheck: service.healthcheck?.test
    }
});

/**
 * Searches for label which satisfies the prefixRegex within the supplied
 * labels
 * @param labels list of labels or undefined when no labels defined
 * @param prefixRegex a regex expression (as string) which when found at
 *      start of label indicates the label which is desired. A partial
 *      prefix will end with a '\.'
 * @returns value of the matched label or undefined if a matching label
 * could not be found or there were no labels
 */
const findLabel = (labels: string[] | undefined, prefixRegex: string) => {
    if (!labels) {
        return undefined;
    }

    // Check whether the prefix is a complete or partial label name (i.e.
    // partial would have a trailing dot) and append '=' to prefix when
    // prefix is complete
    // eslint-disable-next-line no-useless-escape
    if (!prefixRegex.endsWith("\.")) {
        prefixRegex += "=";
    }

    const labelKeyRegex = new RegExp(`^${prefixRegex}.+$`);

    const value = labels.find(label => labelKeyRegex.test(label));
    if (value) {
        return extractLabelValue(value);
    }
    return undefined;
};

const readLabels = (serviceLabels: string[]) => {
    const metadata: Record<string, string> = {};

    for (const [metadataKey, label] of Object.entries(metadataLabelMapping)) {
        const labelValue = findLabel(serviceLabels, label);

        if (labelValue) {
            metadata[metadataKey] = labelValue;
        }
    }

    return metadata;
};

const parseDependsOn: (dependencies?: string[] | DependencySpecificationMap) => string[] = (dependencies?: string[] | DependencySpecificationMap) =>
    dependencies
        ? Array.isArray(dependencies)
            ? dependencies
            : Object.keys(dependencies)
        : [];

const parseRepository = (labels: string[]) => {
    const repository = findLabel(labels, "chs.repository.url");

    return repository && repository !== null
        ? {
            url: repository,
            branch: findLabel(labels, "chs.repository.branch")
        }
        : null;
};

const extractLabelValue = (label: string) => {
    return label.substring(label.indexOf("=") + 1);
};
