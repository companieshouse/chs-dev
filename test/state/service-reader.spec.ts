import { expect } from "@jest/globals";
import { join } from "path";
import { readServices } from "../../src/state/service-reader";
import Service from "../../src/model/Service";

const infraServiceFile = join(
    process.cwd(),
    "test/data/service-reader/infrastructure/mongo.docker-compose.yaml"
);

const simpleServiceFile = join(
    process.cwd(),
    "test/data/service-reader/modules/module-one/simple-service.docker-compose.yaml"
);

const moreComplexServiceFile = join(
    process.cwd(),
    "test/data/service-reader/modules/module-two/more-complex.docker-compose.yaml"
);

const atypicalServiceFile = join(
    process.cwd(),
    "test/data/service-reader/modules/module-three/atypical.docker-compose.yaml"
);

const anotherAtypicalServiceFile = join(
    process.cwd(),
    "test/data/service-reader/modules/module-three/another-atypical.docker-compose.yaml"
);

const repositoryRequiredServiceFile = join(
    process.cwd(),
    "test/data/service-reader/modules/module-one/repository-required.docker-compose.yaml"
);

const deprecatedServiceFile = join(
    process.cwd(),
    "test/data/service-reader/modules/module-three/deprecated-service.docker-compose.yaml"
);

const normaliseLocations = (svcs: Partial<Service>[]) => {
    return svcs.map(svc => {
        svc.source = svc.source?.replace(process.cwd(), ".");
        return svc;
    });
};

describe("readServices", () => {

    it("loads an infrastructure service", () => {
        expect(normaliseLocations(readServices(infraServiceFile))).toMatchSnapshot();
    });

    it("loads a simple service", () => {
        expect(normaliseLocations(readServices(simpleServiceFile))).toMatchSnapshot();
    });

    it("loads a more complicated", () => {
        expect(normaliseLocations(readServices(moreComplexServiceFile))).toMatchSnapshot();
    });

    it("loads atypical with repoContext", () => {
        expect(normaliseLocations(readServices(atypicalServiceFile))).toMatchSnapshot();
    });

    it("loads atypical with entrypoint and outdir", () => {
        expect(normaliseLocations(readServices(anotherAtypicalServiceFile))).toMatchSnapshot();
    });

    it("loads service with repo required", () => {
        expect(normaliseLocations(readServices(repositoryRequiredServiceFile))).toMatchSnapshot();
    });

    it("loads a deprecated service", () => {
        expect(normaliseLocations(readServices(deprecatedServiceFile))).toMatchSnapshot();
    });
});
