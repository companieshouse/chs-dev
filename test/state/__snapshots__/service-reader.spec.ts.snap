// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readServices loads a deprecated service 1`] = `
[
  {
    "builder": "",
    "dependsOn": [
      "mongo",
    ],
    "description": "simple service",
    "metadata": {
      "deprecated": "true",
      "healthcheck": undefined,
    },
    "module": "module-three",
    "name": "simple-service",
    "repository": {
      "branch": undefined,
      "url": "git@github.com/companieshouse/repo.git",
    },
    "source": "./test/data/service-reader/modules/module-three/deprecated-service.docker-compose.yaml",
  },
]
`;

exports[`readServices loads a more complicated 1`] = `
[
  {
    "builder": "java",
    "dependsOn": [
      "simple-service",
    ],
    "description": "more complex service",
    "metadata": {
      "healthcheck": "curl http://localhost/healtcheck",
      "ingressRoute": "Host(\`api.local\`) && (PathPrefix(\`/complex\`))",
      "languageMajorVersion": "17",
    },
    "module": "module-two",
    "name": "more-complex",
    "repository": {
      "branch": "develop",
      "url": "git@github.com/companieshouse/repo2.git",
    },
    "source": "./test/data/service-reader/modules/module-two/more-complex.docker-compose.yaml",
  },
]
`;

exports[`readServices loads a simple service 1`] = `
[
  {
    "builder": "node",
    "dependsOn": [
      "mongo",
    ],
    "description": "simple service",
    "metadata": {
      "healthcheck": undefined,
      "languageMajorVersion": "18",
      "secretsRequired": "true",
    },
    "module": "module-one",
    "name": "simple-service",
    "repository": {
      "branch": undefined,
      "url": "git@github.com/companieshouse/repo.git",
    },
    "source": "./test/data/service-reader/modules/module-one/simple-service.docker-compose.yaml",
  },
]
`;

exports[`readServices loads an infrastructure service 1`] = `
[
  {
    "builder": "",
    "dependsOn": [],
    "description": "primary database storing data",
    "metadata": {
      "healthcheck": "test $$(echo "rs.initiate({_id:'rs',members:[{_id:0,host:\\"mongo:27017\\"}]}).ok || rs.status().ok" | mongo --port 27017 --quiet) -eq 1",
    },
    "module": "infrastructure",
    "name": "mongo",
    "repository": null,
    "source": "./test/data/service-reader/infrastructure/mongo.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "mongo",
    ],
    "description": undefined,
    "metadata": {
      "healthcheck": undefined,
    },
    "module": "infrastructure",
    "name": "execute-mongo-scripts",
    "repository": null,
    "source": "./test/data/service-reader/infrastructure/mongo.docker-compose.yaml",
  },
]
`;

exports[`readServices loads atypical with entrypoint and outdir 1`] = `
[
  {
    "builder": "node",
    "dependsOn": [
      "mongo",
    ],
    "description": "simple service",
    "metadata": {
      "buildOutputDir": "out/",
      "entrypoint": "docker_start.sh",
      "healthcheck": undefined,
      "languageMajorVersion": "20",
    },
    "module": "module-three",
    "name": "another-atypical",
    "repository": {
      "branch": undefined,
      "url": "git@github.com/companieshouse/reapo.git",
    },
    "source": "./test/data/service-reader/modules/module-three/another-atypical.docker-compose.yaml",
  },
]
`;

exports[`readServices loads atypical with repoContext 1`] = `
[
  {
    "builder": "repository",
    "dependsOn": [],
    "description": "atypical service",
    "metadata": {
      "dockerfile": "chs.Dockerfile",
      "healthcheck": undefined,
      "repoContext": "chsDev/",
    },
    "module": "module-three",
    "name": "atypical-service",
    "repository": {
      "branch": undefined,
      "url": "git@github.com/companieshouse/repo3.git",
    },
    "source": "./test/data/service-reader/modules/module-three/atypical.docker-compose.yaml",
  },
]
`;

exports[`readServices loads service with repo required 1`] = `
[
  {
    "builder": "",
    "dependsOn": [],
    "description": undefined,
    "metadata": {
      "dockerfile": "svc1.Dockerfile",
      "healthcheck": undefined,
      "ingressRoute": "Host(\`svc1.chl.local\`)",
      "repositoryRequired": "true",
    },
    "module": "module-one",
    "name": "repo-required",
    "repository": {
      "branch": undefined,
      "url": "git@github.com:companieshouse/repo1.git",
    },
    "source": "./test/data/service-reader/modules/module-one/repository-required.docker-compose.yaml",
  },
]
`;
